### PR TITLE
XML Data support - phase one

### PR DESCRIPTION
**Pull request by baughj**
(@kojasou was kind enough to point out I never did this:
I have read and agree to the terms of the Contributor Agreement published at https://github.com/hybrasyl/server/blob/master/CONTRIBUTING.md . )

# A New Data Model for Hybrasyl

This PR begins the introduction a completely new data model for Hybrasyl, using XML for static (world) data and Redis for everything else (mail/board posts, user data, etc).

## Why am I proposing this?

Currently the MySQL schema is fairly messy and we maintain two ORM interfaces to the data - one in Entity Framework for C#, and one in Rails using ActiveRecord. Any update to new tables (generally done in Ealagad) means making concomitant updates to EF as well. Over time this has gotten harder to manage as implementing a feature in Hybrasyl basically means also implementing all the Rails to allow objects to be edited/changed/etc. This has slowed down development on a number of complex types (most notably castables) and it requires potential contributors to the project to either hope someone will write an ActiveAdmin interface or requires them to know both C# and Ruby and have a willingess to also create an admin interface.

It is my hope that by using a well-defined XML format with validating schema, we can provide a simple GUI interface to editing static data (and indeed, I have tried to make the XML format as human friendly as possible - intrepid souls can simply open a text editor). I believe that using XML has a number of advantages: load time (on the order of 500ms for 800+ maps), automatic validation of input, ease of extendability & schema updates. The EF interface in Hybrasyl is an afterthought and is not represented by tightly coupled objects (nor are they easily-serializable POCO), so there is a number of steps required to actually load data from SQL into Hybrasyl. 

Further: MySQL for .NET sucks. It often lags behind new EF versions, has varying levels of support for features, and is extremely challenging to debug / investigate. I myself have lost hours of time to this. 

Moving to XML and Redis will significantly reduce the complexity of the server by using two extremely well known, high performance, tested interfaces (e.g. C#'s built in XML support and StackExchange's Redis client) as opposed to something hacked in in the short term to remove the previous habit of executing direct SQL.

This PR is the first step, and introduces all the new XML schema as embedded resources in a new directory. I figured it would be useful to break out the XML data changes into several steps - the first step, introducing the schema and related XSDs for discussion, the second step, actually breaking away from MySQL for static data (e.g. using XML for world data entirely) and the last step, using Redis for everything else.

Although users would still need to install Redis on Windows, this is easily done (with a port maintained by Microsoft) and much easier to set up and use than MySQL. The other obvious impact is that Ealagad would be rendered obsolete, and the current password reset functionality for  players would need to be reimplemented using Redis. We would also need to prioritize development of the editor as a window into all of this data, but honestly, I feel that a GUI editor is really where we want/need to go anyway.

Also included are example files that can be parsed using the schema files either using XML support in C# or with `xmllint` (part of libxml2).

There are some implications of using XML, however:

##### Almost all numeric IDs are deprecated except where required by the client (map numbers being the notable exception).

This means that a number of things will be indexed by name - and name will be enforced to be unique in the system. This is already the case in a number of places (e.g. map names cannot be identical). As part of step two, there will be a significant refactoring of Hybrasyl's object indexes and storage (Wren had a great point when he said we should implement an object manager).

##### NPCs are collapsed to Map level - they aren't separate entities

This introduces a restriction that there can't be more than one NPC on a map named the same thing - which I think is probably OK. It does however mean that there can be any number of instances of the same-named NPC (with different settings) on different maps.

##### A number of simple effects & scripting hooks are expressible in XML

Signposts / boards can have scripts attached to them (as can most clickables) via the <scriptname> element; items can create simple effects based on XML directives rather than requiring Python (teleporting being an obvious example). In this way, we should be able to lower the bar / simplify basic cases for certain items & effects vs having to use Python to express absolutely every effect in the game.

Comments and thoughts are very welcome.

cc: @luke-segars @fluffguck @kojasou @Blackbaud-PeteLukich

### CODE REVIEW

**Comment by baughj** [comment]
@luke-segars @kojasou any thoughts before I merge this?

**Comment by kojasou** [comment]
My only concern was the NPC restriction. I'm unable to think of any instances where the inability to have multiple same-named NPCs on a single map would be an issue (other than Fish, which is very easily resolvable by implementing them simply as a monster rather than NPC), so I think this is fine. Everything else looks good to go.

**Comment by Fluffguck** [comment]
Arena will need to be adjusted. Only one I can think of off-hand.

**Comment by kojasou** [comment]
Actually, each NPC in arena is named differently: Bastill, Bastilla, Bastille, Bastilli, ColiseumTir, ColiseumTor, ColiseumTur, and ColiseumTyr.

**Comment by Fluffguck** [comment]
I stand corrected. Also that is hilarious.

File: hybrasyl/XML/Examples/items/Blouse.xml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    <appearance sprite="115" equip="-1" display="1" bodystyle="transparent"/>
+    <stackable max="1"/>
+    <physical value="0" weight="0" durability="0" perishable="false" vendorable="false" bound="false"/>
+    <equipment slot="armor" unique="false"/>
+    <stateffects>
+      <base str="0" con="0" int="0" wis="0" dex="0" hp="0" mp="0"/>
+      <combat hit="0" dmg="0" ac="0" regen="0" mr="0"/>
+      <elemental defense="0"/>
+    </stateffects>
+    <flags>depositable enchantable exchangeable</flags>
```
Comment (luke-segars) [inline hybrasyl/XML/Examples/items/Blouse.xml]: Nit but do you think it's better to have flags as a set of strings vs
    <flags>
        <depositable />
        <enchantable />
        <exchangeable />
    </flags>

We could validate flags that way; don't think it matters too much though.

**Comment by luke-segars** [comment]
This look great. Added one small comment.

Do you think there's anything important from a validation perspective that would should add to continuous integration in the future? Maybe verify that everything in `hybrasyl/XML/Examples/` passes xmllint? If these are the only examples we expect then we can probably go without; if others will be editing and committing then it may make sense to add something.

Good stuff!

File: hybrasyl/XML/Examples/items/Blouse.xml (hunk 1 above)
Comment (baughj) [inline hybrasyl/XML/Examples/items/Blouse.xml]: Flags are actually validated; the content of  the flags element is a list, required to be one of the defined flags in `hyb:ItemFlags` from `hybrasylTypes.xsd`.

Comment (baughj) [inline hybrasyl/XML/Examples/items/Blouse.xml]: Revisiting this now that I'm working with actually deserializing items.

**Comment by baughj** [comment]
Starting to work on this again finally, and I need this merged.

**Pull request merged**